File: src/core/Math2D.ts
/**
 * Math2D - allocation-free 2D math helpers.
 *
 * The whole simulation runs on a flat XZ plane (top-down world space) with a
 * separate scalar `y` used only for bullet height / eye height. This mirrors
 * Unity's world space so ports map 1:1: our (x, y) is Unity's (x, z).
 *
 * Every function here is pure and allocation-free where possible - these are
 * called thousands of times per frame by ballistics and AI.
 */

export interface Vec2 {
  x: number;
  y: number;
}

export const TAU = Math.PI * 2;
export const DEG2RAD = Math.PI / 180;
export const RAD2DEG = 180 / Math.PI;

export function vec2(x = 0, y = 0): Vec2 {
  return { x, y };
}

export function clamp(v: number, min: number, max: number): number {
  return v < min ? min : v > max ? max : v;
}

export function clamp01(v: number): number {
  return v < 0 ? 0 : v > 1 ? 1 : v;
}

export function lerp(a: number, b: number, t: number): number {
  return a + (b - a) * t;
}

export function inverseLerp(a: number, b: number, v: number): number {
  return a === b ? 0 : clamp01((v - a) / (b - a));
}

/** Frame-rate independent exponential smoothing. `rate` = how much of the gap closes per second. */
export function damp(current: number, target: number, rate: number, dt: number): number {
  return lerp(current, target, 1 - Math.exp(-rate * dt));
}

export function moveTowards(current: number, target: number, maxDelta: number): number {
  const d = target - current;
  if (Math.abs(d) <= maxDelta) return target;
  return current + Math.sign(d) * maxDelta;
}

export function distance(ax: number, ay: number, bx: number, by: number): number {
  const dx = bx - ax;
  const dy = by - ay;
  return Math.sqrt(dx * dx + dy * dy);
}

export function distanceSq(ax: number, ay: number, bx: number, by: number): number {
  const dx = bx - ax;
  const dy = by - ay;
  return dx * dx + dy * dy;
}

/** Wrap an angle into (-PI, PI]. */
export function wrapAngle(a: number): number {
  a = (a + Math.PI) % TAU;
  if (a < 0) a += TAU;
  return a - Math.PI;
}

/** Shortest signed delta from `from` to `to`, in radians. */
export function angleDelta(from: number, to: number): number {
  return wrapAngle(to - from);
}

/** Rotate `current` towards `target` by at most `maxDelta` radians. */
export function rotateTowards(current: number, target: number, maxDelta: number): number {
  const d = angleDelta(current, target);
  if (Math.abs(d) <= maxDelta) return wrapAngle(target);
  return wrapAngle(current + Math.sign(d) * maxDelta);
}

export function angleLerp(a: number, b: number, t: number): number {
  return wrapAngle(a + angleDelta(a, b) * t);
}

/** Smooth Hermite interpolation, used for falloff curves. */
export function smoothstep(edge0: number, edge1: number, x: number): number {
  const t = clamp01((x - edge0) / (edge1 - edge0));
  return t * t * (3 - 2 * t);
}

/** Remaps v from [inMin,inMax] into [outMin,outMax] with clamping. */
export function remap(v: number, inMin: number, inMax: number, outMin: number, outMax: number): number {
  return lerp(outMin, outMax, inverseLerp(inMin, inMax, v));
}

/** Axis-aligned circle vs circle overlap. */
export function circlesOverlap(ax: number, ay: number, ar: number, bx: number, by: number, br: number): boolean {
  const r = ar + br;
  return distanceSq(ax, ay, bx, by) <= r * r;
}

/**
 * Squared distance from point P to segment AB. Used for muzzle-line hit tests
 * and "is this AI near the patrol path" checks without allocating.
 */
export function pointSegmentDistSq(px: number, py: number, ax: number, ay: number, bx: number, by: number): number {
  const abx = bx - ax;
  const aby = by - ay;
  const apx = px - ax;
  const apy = py - ay;
  const lenSq = abx * abx + aby * aby;
  const t = lenSq > 0 ? clamp01((apx * abx + apy * aby) / lenSq) : 0;
  const cx = ax + abx * t;
  const cy = ay + aby * t;
  return distanceSq(px, py, cx, cy);
}

/** Normalizes a vector in place; returns the original length. */
export function normalize(v: Vec2): number {
  const len = Math.sqrt(v.x * v.x + v.y * v.y);
  if (len > 1e-6) {
    v.x /= len;
    v.y /= len;
  }
  return len;
}

/** Clamps a vector's magnitude to `max` in place. */
export function clampMagnitude(v: Vec2, max: number): void {
  const lenSq = v.x * v.x + v.y * v.y;
  if (lenSq > max * max) {
    const s = max / Math.sqrt(lenSq);
    v.x *= s;
    v.y *= s;
  }
}

File: unity/Assets/Greyzone/Simulation/Core/Math2D.cs
using System;
using System.Runtime.CompilerServices;

namespace Greyzone.Simulation.Core
{
    /// <summary>
    /// Allocation-free 2D maths shared by the whole simulation.
    /// </summary>
    /// <remarks>
    /// The simulation runs on a flat plane with a separate scalar height used
    /// only for bullet and eye elevation. Our (x, y) is Unity's (x, z), so a
    /// world position converts with <c>new Vector3(p.X, height, p.Y)</c> and
    /// nothing here needs to know that.
    ///
    /// Every method is pure and allocation-free: ballistics and AI call these
    /// thousands of times per tick.
    /// </remarks>
    public static class Math2D
    {
        public const float Tau = (float)(Math.PI * 2.0);
        public const float Deg2Rad = (float)(Math.PI / 180.0);
        public const float Rad2Deg = (float)(180.0 / Math.PI);
        public const float Pi = (float)Math.PI;

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static float Clamp(float v, float min, float max) => v < min ? min : (v > max ? max : v);

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static int Clamp(int v, int min, int max) => v < min ? min : (v > max ? max : v);

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static float Clamp01(float v) => v < 0f ? 0f : (v > 1f ? 1f : v);

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static float Lerp(float a, float b, float t) => a + (b - a) * t;

        public static float InverseLerp(float a, float b, float v)
            => a == b ? 0f : Clamp01((v - a) / (b - a));

        /// <summary>
        /// Frame-rate independent exponential smoothing: <paramref name="rate"/>
        /// is the fraction of the remaining gap closed per second.
        /// </summary>
        /// <remarks>
        /// Using this instead of a raw <c>Lerp(current, target, 0.1f)</c> is what
        /// keeps camera and aim smoothing identical at 60 and 120 FPS.
        /// </remarks>
        public static float Damp(float current, float target, float rate, float dt)
            => Lerp(current, target, 1f - (float)Math.Exp(-rate * dt));

        public static float MoveTowards(float current, float target, float maxDelta)
        {
            float d = target - current;
            if (Math.Abs(d) <= maxDelta) return target;
            return current + Math.Sign(d) * maxDelta;
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static float Distance(float ax, float ay, float bx, float by)
        {
            float dx = bx - ax;
            float dy = by - ay;
            return (float)Math.Sqrt(dx * dx + dy * dy);
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static float DistanceSq(float ax, float ay, float bx, float by)
        {
            float dx = bx - ax;
            float dy = by - ay;
            return dx * dx + dy * dy;
        }

        /// <summary>Wraps an angle into (-PI, PI].</summary>
        public static float WrapAngle(float a)
        {
            a = (a + Pi) % Tau;
            if (a < 0f) a += Tau;
            return a - Pi;
        }

        /// <summary>Shortest signed delta from one heading to another, in radians.</summary>
        public static float AngleDelta(float from, float to) => WrapAngle(to - from);

        /// <summary>Rotates towards a heading by at most <paramref name="maxDelta"/> radians.</summary>
        public static float RotateTowards(float current, float target, float maxDelta)
        {
            float d = AngleDelta(current, target);
            if (Math.Abs(d) <= maxDelta) return WrapAngle(target);
            return WrapAngle(current + Math.Sign(d) * maxDelta);
        }

        public static float AngleLerp(float a, float b, float t) => WrapAngle(a + AngleDelta(a, b) * t);

        /// <summary>Smooth Hermite interpolation, used for falloff curves.</summary>
        public static float Smoothstep(float edge0, float edge1, float x)
        {
            float t = Clamp01((x - edge0) / (edge1 - edge0));
            return t * t * (3f - 2f * t);
        }

        public static float Remap(float v, float inMin, float inMax, float outMin, float outMax)
            => Lerp(outMin, outMax, InverseLerp(inMin, inMax, v));

        public static bool CirclesOverlap(float ax, float ay, float ar, float bx, float by, float br)
        {
            float r = ar + br;
            return DistanceSq(ax, ay, bx, by) <= r * r;
        }

        /// <summary>
        /// Squared distance from a point to a segment.
        /// </summary>
        /// <remarks>
        /// This is the hot path for bullet-versus-actor tests: a round's travel
        /// over one tick is a segment, and "did it pass within body radius" is
        /// exactly this comparison. Clamping to the segment (rather than the
        /// infinite line) is what stops a shot from registering behind you.
        /// </remarks>
        public static float PointSegmentDistSq(float px, float py, float ax, float ay, float bx, float by)
        {
            float abx = bx - ax;
            float aby = by - ay;
            float apx = px - ax;
            float apy = py - ay;
            float lenSq = abx * abx + aby * aby;
            float t = lenSq > 0f ? Clamp01((apx * abx + apy * aby) / lenSq) : 0f;
            float cx = ax + abx * t;
            float cy = ay + aby * t;
            return DistanceSq(px, py, cx, cy);
        }

        /// <summary>Normalises in place and returns the original length.</summary>
        public static float Normalize(ref float x, ref float y)
        {
            float len = (float)Math.Sqrt(x * x + y * y);
            if (len > 1e-6f)
            {
                x /= len;
                y /= len;
            }
            return len;
        }

        /// <summary>Clamps a vector's magnitude in place.</summary>
        public static void ClampMagnitude(ref float x, ref float y, float max)
        {
            float lenSq = x * x + y * y;
            if (lenSq > max * max)
            {
                float s = max / (float)Math.Sqrt(lenSq);
                x *= s;
                y *= s;
            }
        }
    }

    /// <summary>
    /// Plain 2D point on the simulation plane.
    /// </summary>
    /// <remarks>
    /// A struct rather than a class so paths, waypoints and spawn lists do not
    /// each become a few hundred heap allocations per raid.
    /// </remarks>
    public readonly struct Vec2 : IEquatable<Vec2>
    {
        public readonly float X;
        public readonly float Y;

        public Vec2(float x, float y)
        {
            X = x;
            Y = y;
        }

        public static Vec2 operator +(Vec2 a, Vec2 b) => new Vec2(a.X + b.X, a.Y + b.Y);
        public static Vec2 operator -(Vec2 a, Vec2 b) => new Vec2(a.X - b.X, a.Y - b.Y);
        public static Vec2 operator *(Vec2 a, float s) => new Vec2(a.X * s, a.Y * s);

        public float Length => (float)Math.Sqrt(X * X + Y * Y);

        public bool Equals(Vec2 other) => X.Equals(other.X) && Y.Equals(other.Y);

        public override bool Equals(object obj) => obj is Vec2 other && Equals(other);

        public override int GetHashCode()
        {
            unchecked
            {
                return (X.GetHashCode() * 397) ^ Y.GetHashCode();
            }
        }

        public override string ToString() => $"({X:0.##}, {Y:0.##})";
    }
}
